src/wallet/api.rs

```diff
@@ -1255,63 +1260,97 @@ impl Wallet {
         let target_weight = Weight::from_vb_unwrap(
             (Amount::SIZE + VarInt::from(P2WPKH_SPK_SIZE).size() + P2WPKH_SPK_SIZE) as u64,
         );
-        let avg_output_weight = (change_weight.to_wu() + (target_weight.to_wu())) / 2;
+        let avg_output_weight = (change_weight.to_wu() + target_weight.to_wu()) / 2;
         let avg_input_weight = unspents
             .iter()
             .map(|(_, spend_info)| {
                 // Base weight of the input (OutPoint + sequence + empty script_sig)
-                let base_weight = 32 + 4 + 4 + 1; // 32 bytes for OutPoint, 4 bytes for sequence,4 bytes for vout, 1 byte for empty script_sig
+                let input_base_weight = 32 + 4 + 4 + 1; // OutPoint + sequence + vout + empty script_sig
                 let witness_weight = spend_info.estimate_witness_size();
-                base_weight + witness_weight as u64
+                input_base_weight + witness_weight as u64
             })
             .sum::<u64>()
             / unspents.len() as u64;
 
-        // Note: Consider more sophisticated grouping policies in the future
-        // Group UTXOs by address
-        // TODO: Change this to HashMap<String, Vec<(ListUnspentResultEntry, UTXOSpendInfo)>>
-        let mut address_groups: Vec<(String, Vec<(ListUnspentResultEntry, UTXOSpendInfo)>)> =
-            Vec::new();
+        let target_amount = amount.to_sat();
 
-        for (utxo, spend_info) in &unspents {
+        let mut address_groups: HashMap<String, Vec<(ListUnspentResultEntry, UTXOSpendInfo)>> =
+            HashMap::new();
+        for (utxo, spend_info) in unspents {
             let address_str = utxo
                 .address
                 .as_ref()
                 .map(|addr| addr.clone().assume_checked().to_string())
                 .unwrap_or_else(|| format!("script_{}", utxo.script_pub_key));
+            address_groups
+                .entry(address_str)
+                .or_default()
+                .push((utxo, spend_info));
+        }
+
+        let (mut grouped_addresses, single_addresses): (Vec<_>, Vec<_>) = address_groups
+            .into_values()
+            .partition(|group| group.len() > 1);
+
+        // Sort by value
+        grouped_addresses
+            .sort_by_key(|group| group.iter().map(|(u, _)| u.amount.to_sat()).sum::<u64>());
+
+        // Mojo's Ultra-Compressed: Single loop with let binding
```

Comment:
Nicely done. And thanks for the credit. I would love to keep it this way, but for the sake of clarity, we should say something like "Check if the grouped utxos match the target".